README.md
# Classes
## ```class HoloeyeSLM```
### Description
This class has the main role in commmunication with the Holoeye device. This class inherits the ```SSHClient``` class from ```paramiko``` package and has all the properties from the same package. 
### Methods
* ```HoloeyeSLM() ``` aka constructor
    * Description 

        Initializes the connection to your Holoeye SLM device
    * Parameters: 
        > *```parameter=default```          description*
        * ```logging=False ```      Enables the logs from your HOLOEYE device.

        * ```RSA_Keys=None ```      Path to your prefered RSA keys.  

        * ```width=1920```          Width of your HOLOEYE device.  

        * ```height=1080```         Width of your HOLOEYE device.  

        * ```min=0```               Minimum intensity withing the Holoeye device supporting range  

        * ```max=255```             Maximum intensity withing the Holoeye device supporting range 

        * ```hostname=10.10.70.1``` Your device hostname/IP address 

        * ```port=22```             Your device listening port  

        * ```username=root```       Your device end username

        * ```password=''```         Your device end password 

* ```flush_RSA_Keys()```
    * Description 

        Deletes the current RSA keys.


* ```prepare_connect()```
    * Description 
    
        Reinitialize and connect to the device with new changes in device parameters. e.g. changing the device IP address.

* ```connect()```
    * Description 
    
        Connects to the device.

* ```changeIP(new_IP,mask)```
    * Description
        
        Changes the device IP configuration to the given IP and network submask and closes the connection. It is recommentded to set your IP addresses with default submask. After changing the device IP configuration, you need to **re-plug** the device to your computer **without turning off** the Holoeye device.
    * Parameters
        * ```new_IP='10.10.70.2'``` Desisered IPv4 address for the device
        * ```mask='255.0.0.0'```    Desisered mask for the device

* ```diconnectHDMI()```
    * Description 
    
        Disconnects the HDMI port access from the Holoeye device.

* ```sendImage(FILE_PATH)```
    * Description 
    
        Sends the image at given path to the Holoeye device output.
    * Parameters    
        > *```parameter```          description*
        * ```FILE_PATH```           path to image file. * image file must be a bitmap of 2-D array with dimentions equal to ```width``` and ```height``` attributes of the class as well as pixel values withing the class ```min``` and ```max``` parameters.

* ```sendData(data)```
    * Description 
    
        Sends the given array values to the Holoeye device output.
    * Parameters    
        > *```parameter```          description*
        * ```data```                numput 2-D array with dimentions equal to ```width``` and ```height``` attributes of the class as well as pixel values withing the class ```min``` and ```max``` parameters.
# Device Driver Installation and Configuration
## Windows
### Installing RNDIS Driver on Windows 
1. Connect the HoloeyeSLM USB-OTG port to your computer. *Do not connect HDMI to your computer*
2. Open DeviceManager on your computer. 
    
    > To open DeviceManager on your pc, open run(win+R) and type ```devmgmt.msc``` and press Enter button.
    
    If you don't have the RNDIS driver installed on your computer for that port, you are supposed to following:  
 
    ![RNDIS](https://i.ibb.co/kDqpwzD/RNDIS1.png) 
 
3. Right click on RNDIS and choose ```Update Driver```
    
    ![RNDIS](https://i.ibb.co/SsnBTC2/RNDIS2.png)
4. Click on ```Browse my computer for drivers```
    ![RNDIS](https://i.ibb.co/k6Qd067/RNDIS3.png)
5. Click on ``` Let me pick from a list of ... ``` and click on ``` Next ```
    ![RNDIS](https://i.ibb.co/m4VQkK3/RNDIS4.png)
6. From the list choose ```Network adapters``` as your decive type and click on ``` Next ```
    ![RNDIS](https://i.ibb.co/hsN5Fg7/RNDIS5.png)
7. From the list of Manufactures, choose ``` Microsoft ``` and from Model list choose USB RNDIS Adapter and click on ``` Next ``` 
    ![RNDIS](https://i.ibb.co/TRMx6mm/RNDIS6.png)
8. At this step you would have a warning about the possibility of uncompatibality of the chose driver with your device .(**Disclaimer**: I verified to compatibility with with **Holoeye Pluto-2**). Choose ```yes``` for the poping warning dialog.  
    ![RNDIS](https://i.ibb.co/cQqxc37/RNDIS7.png)
9. Now you have installed the RNDIS network driver.Now, you can close the window and proceed with Network Configuration.
    ![RNDIS](https://i.ibb.co/SdRpPzD/RNDIS8.png)
10. Currently, you should be able to see your RNDIS network adapter in thus section
    > Number of RNDIS adapter is equal to number of the devices connected to your computer.
    ![RNDIS](https://i.ibb.co/kJT4gwm/RNDIS9.png)

### Configuring RNDIS network address and gateway

1. Open ```Control Panel```
    ![RNDIS](https://i.ibb.co/17R33LF/RNDIS10.png)
2. Click on ```Network and Internet```
    ![RNDIS](https://i.ibb.co/pwrNMVG/RNDIS11.png)
3. Click on ```Network and Sharing Center```
    ![RNDIS](https://i.ibb.co/FJggyQG/RNDIS12.png)
4. Click on ```Change adapter setting```
    ![RNDIS](https://i.ibb.co/GFcHz7H/RNDIS13.png)
5. Right click on network adapter which corresponds to your desiered RNDIS adapter and click on ```Properties```.
    ![RNDIS](https://i.ibb.co/BLNQVcR/RNDIS14.png)
6. From list items choose ```Internet Portocol Version 4(TCP/IPv4)``` and click on ```Properties```.
    ![RNDIS](https://i.ibb.co/Gs1b0kH/RNDIS15.png)

7. Click on ```Use the following IP address```. Enter IP address regarding your device assigned IP. The default IP configuration for device is ```10.10.70.1/8```. Thus you may use any IP address which does not conflict with your other RNDIS network configurations and devices and click on ```Ok```. E.g : 
    ``` 
    # device IP on RNDIS adapter 1 ->  10.10.70.1
    IP address : 10.10.70.254 # RNDIS adapter 1 IP addr 
    Subnet mask : 255.0.0.0   # RNDIS adapter 1 Subnet mask 
    
    # device IP on RNDIS adapter 2 ->  10.10.70.2
    IP address : 10.10.70.253 # RNDIS adapter 2 IP addr 
    Subnet mask : 255.0.0.0   # RNDIS adapter 2 Subnet mask 

    ...
    ```
    ![RNDIS](https://i.ibb.co/0yV3smD/RNDIS16.png)
8. Click on ```Close``` and exit. Now you are set up with your devices.
    ![RNDIS](https://i.ibb.co/60PsR7b/RNDIS17.png)

## Linux 
You may use [```ifconfig```](https://man7.org/linux/man-pages/man8/ifconfig.8.html) for configuing your device. In order to assign the IP address, consider the following example: 
``` 
    # device IP on usb0 ->  10.10.70.1
    IP address : 10.10.70.254 # usb0 adapter  IP addr 
    Subnet mask : 255.0.0.0   # usb0 adapter  Subnet mask 
    
    # device IP on usb1 adapter  ->  10.10.70.2
    IP address : 10.10.70.253 # usb1 adapter IP addr 
    Subnet mask : 255.0.0.0   # usb1 adapter Subnet mask 
    ...
```
## MACOS
You may use [HoRNDIS](https://www.joshuawise.com/horndis) for configuring your device connection. Make sure you preserver the address assignment for your different devices. Please refer to linux instruction or step 7 of windows instrction for a more clear example. 


# CHANGELOG
Version 1.0.0 (2021-11-12)
------------------------

* Changing device address API added.
* numpy array support
* Windows configuration support added.
* Improvement on documention

Version 0.1.0 (2021-10-15)
------------------------

* First release after the conversion to the perf module and move to GitHub
* Included RNDIS and SSH API.

# CONTRIBUTING
HOLOEYE-EMBEDDED is As an open source project, contributions are welcomed of many forms.
Examples of contributions include:
* Code patches
* Suggestion or Implemention of new APIs
* Documentation improvements
* Bug reports and patch reviews

CHANGELOG.RST
Version 1.0.1 (2021-11-12)
------------------------
* Minor Bug fix

Version 1.0.0 (2021-11-12)
------------------------

* Changing device address API added.
* numpy array support
* Windows configuration support added.
* Improvement on documention

Version 0.1.0 (2021-10-15)
------------------------

* First release after the conversion to the perf module and move to GitHub
* Included RNDIS and SSH API.


test/test.py
import paramiko,time
if __name__ == '__main__':
    host = "10.10.70.1"
    port = 22
    username = "root"
    password = ""

    # sshClient = paramiko.SSHClient()
    # sshClient.connect(hostname=host, username=username, password=password)
    # channel = sshClient.get_transport().open_session()
    # channel.get_pty()
    # channel.invoke_shell()

    # while True:
    #     command = input('$ ')
    #     if command == 'exit':
    #         break

    #     channel.send(command + "\n")

    #     while True:
    #         if channel.recv_ready():
    #             output = channel.recv(1024)
    #             print(output)
    #         else:
    #             time.sleep(0.5)
    #             if not(channel.recv_ready()):
    #                 break

    # sshClient.close()

########################################

    # command = "ls"

    # ssh = paramiko.SSHClient()
    # ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    # ssh.connect(host, port, username, password)
    # while True:
    #     command = input('\n$> ')
    #     stdin, stdout, stderr = ssh.exec_command(command+';\n')
    #     lines = stdout.read()
    #     print(lines.decode())

#########################################

    ssh = paramiko.SSHClient()
    ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    ssh.connect(host, username=username, password=password, port=22)

    channel = ssh.invoke_shell()

    out = channel.recv(9999)

    channel.send('cd /mnt\n')
    channel.send('ls\n')

    while not channel.recv_ready():
        pass
    out = channel.recv(9999)
    print(out.decode("ascii"))

holoeye_embedded/__meta__.py
# -*- coding: utf-8 -*-

from __future__ import unicode_literals

__all__ = [
    "__packagename__",
    "__version__",
    "__summary__",
    "__url__",
    "__author__",
    "__email__",
    "__license__"
]

# The name of the package 
__packagename__ = "holoeye_embedded"


__version__ = "0.1.0"

# A one-line summary of what the package does. It's like the first line of a doc string. 
__summary__ = "An alternative interface to official API of the Holoeye PLUTO-2 Spatial Light Modulator in order to control multiple devices through RNDIS network rather than classic HDMI, though you don't need to buy multiple vide cards for communicating to all your beloved SLMs."

# The URL of the package's home page 
__url__ = "https://github.com/ehsan2754/HOLOEYE-PLUTO-2-EMBEDDED"

# The name of the author 
__author__ = "Ehsan Shaghaei"

# The email address of the author. (PEP 241 mentions that this might be used as a unique key in some catalog of packages, but I don't know if it actually is.) 
__email__ = "e.shaghaei@innopolis.university"

# PEP 241 says to put the name of the license here, but I don't think that's a good idea. The name doesn't identify a specific license. There are several Zope licenses and several PSF licenses. I recommend putting the URL of the license. 
__license__ = "MIT"


holoeye_embedded/holoeye_slm.py
import threading
import numpy as np
import imageio
import os
import paramiko
from paramiko import SSHClient
from scp import SCPClient


class HoloeyeSLM (SSHClient):
    class Commands:
        PWD = 'pwd'
        GO_HOME = "cd ~"
        SET_LIBRARY = "export LD_LIBRARY_PATH=/mnt"
        DISABLE_HDMI = "/mnt/ControlExmpl -a"
        SHOW_IMAGE = "/mnt/ControlExmpl -o ~/tmp.bmp"

    invalid_input_err = "INVALID input: the size of input image/array must be({0},{1}) and the indices/pixels values must be a number between ({2},{3})"
    CACHE_PATH = "./cache"

    '''
    Initializes and establishes the connection to the device
    '''

    def __init__(self, host='10.10.70.1', port=22, username='root', password='', width=1920, height=1080, min=0, max=255, logging=True):
        self.logging = logging
        if(not os.path.exists(self.CACHE_PATH)):
            os.mkdir(self.CACHE_PATH)
        self.width = width
        self.height = height
        self.min = min
        self.max = max
        self.invalid_array_err = self.invalid_input_err.format(
            height, width, min, max)
        self.hostname = host
        self.port = port
        self.username = username
        self.password = password
        super().__init__()
        self.set_missing_host_key_policy(paramiko.AutoAddPolicy())
        self.connect()
        self.channel = self.invoke_shell()
        self._go_home()
        self._set_library_path()
        self.diconnectHDMI()
        self.logger = threading.Thread(target=self.__log,daemon=True,name='HoloeyeSLM {0} Logger'.format(self.hostname))
        # self.logger.start()

    def __log(self):
        pass
        # while (self.logging):
        #     print(">>LOG : ")
        #     while not self.channel.recv_ready():
        #         pass
        #     out = self.channel.recv(9999)
        #     print(out)

    '''
    gets the current session directory
    '''

    def _pwd(self):
        stdin, stdout, stderr = self.exec_command(self.Commands.PWD)
        lines = stdout.readlines()
        return lines

    '''
    Setups the current session directory to home
    '''

    def _go_home(self):
        self.channel.send(self.Commands.GO_HOME+'\n')
        # while not self.channel.recv_ready():
        #         pass
        # out = self.channel.recv(9999)
        # print(out.decode())


    '''
    Setups the libraries path in SLM
    '''

    def _set_library_path(self):
        self.channel.send(self.Commands.SET_LIBRARY+'\n')
        # while not self.channel.recv_ready():
        #         pass
        # out = self.channel.recv(9999)
        # print(out.decode())

    '''
    Runs the show command on slm
    '''

    def _show_image(self):
        self.channel.send(self.Commands.SHOW_IMAGE+'\n')
        # while not self.channel.recv_ready():
        #         pass
        # out = self.channel.recv(9999)
        # print(out.decode())

    '''
    Validates the array whether it is a legitimate input.
    '''

    def _validateArray(self, array):
        if (array.shape == (self.height, self.width)) and (array.max() <= self.max) and (array.min() >= self.min):
            return array
        raise Exception(self.invalid_input_err)

    '''
    Validates the input image path whether it is a legitimate input. And returns the array in case it's a correct format
    '''

    def _validateImage(self, path):
        im = imageio.imread(path)
        return self._validateArray(im)

    '''Saves the image in the temporary path'''

    def _saveImage(self, array):
        try:
            imageio.imwrite(self.CACHE_PATH+'/tmp.bmp', array)
        except Exception as ex:
            print(ex)

    '''Removes the image in the temporary path'''

    def _removeImage(self, array):
        try:
            if os.path.exists(self.CACHE_PATH+'/tmp.bmp'):
                os.remove(self.CACHE_PATH+'/tmp.bmp')
        except Exception as ex:
            print(ex)

    '''
    Establishes connection to the Holoeyes SLM device
    '''

    def connect(self):
        return super().connect(self.hostname, port=self.port, username=self.username, password=self.password)

    '''
    Disoconnects the HDMI from the HoloeyeSLM device
    '''

    def diconnectHDMI(self):
        self.channel.send(self.Commands.DISABLE_HDMI+'\n')
        # while not self.channel.recv_ready():
        #         pass
        # out = self.channel.recv(9999)
        # print(out.decode())

    '''
    Sends an image at a given path to HoloeyeSLM
    '''

    def sendImage(self, FILE_PATH):
        try:
            im = imageio.imread(FILE_PATH)
            array = self._validateImage(FILE_PATH)
            self._saveImage(array)
            # TODO : SEND IMAGE TO HOME DIRECTORY
            with SCPClient(self.get_transport()) as scp:
                scp.put(self.CACHE_PATH+'/tmp.bmp',remote_path='~')
                scp.close()
            self._show_image()
            # TODO : RUN SHOW COMMAND ON

            # self._removeImage()
        except Exception as ex:
            raise ex


holoeye_embedded/__init__.py
import threading
import numpy as np
import imageio
import os
import paramiko
from paramiko import SSHClient
from scp import SCPClient


class HoloeyeSLM (SSHClient):
    class Commands:
        PWD = 'pwd'
        GO_HOME = "cd ~"
        SET_LIBRARY = "export LD_LIBRARY_PATH=/mnt"
        DISABLE_HDMI = "/mnt/ControlExmpl -a"
        SHOW_IMAGE = "/mnt/ControlExmpl -o ~/tmp.bmp"
        CHANGE_IP = "ifconfig usb0 {0} {1}"

    invalid_input_err = "INVALID input: the size of input image/array must be({0},{1}) and the indices/pixels values must be a number between ({2},{3})"
    CACHE_PATH = "./cache"

    '''
    Initializes and establishes the connection to the device
    '''
    def flush_RSA_Keys(self):
        if self.RSA_Keys is None:
            # try the user's .ssh key file, and mask exceptions
            self.RSA_Keys = os.path.expanduser("~/.ssh/known_hosts")
        try:
            fb = open(self.RSA_Keys,'wb')
            fb.write(b'')
            fb.close()
        except IOError:
            pass

    def __init__(self, host='10.10.70.1', port=22, username='root', password='', width=1920, height=1080, min=0, max=255, logging=False,RSA_Keys=None):
        self.logging = logging
        self.RSA_Keys = RSA_Keys
        if(not os.path.exists(self.CACHE_PATH)):
            os.mkdir(self.CACHE_PATH)
        self.width = width
        self.height = height
        self.min = min
        self.max = max
        self.invalid_array_err = self.invalid_input_err.format(
            height, width, min, max)
        self.hostname = host
        self.port = port
        self.username = username
        self.password = password
        self.prepare_connect()

    def prepare_connect(self):
        self.flush_RSA_Keys()
        super().__init__()
        self.set_missing_host_key_policy(paramiko.AutoAddPolicy())
        self.connect()
        self.channel = self.invoke_shell()
        self._go_home()
        self._set_library_path()
        self.diconnectHDMI()
        self.logger = threading.Thread(target=self.__log,daemon=True,name='HoloeyeSLM {0} Logger'.format(self.hostname))
        self.logger.start()
    
    def __log(self):
        while (self.logging):
            print("Holoeye SLM Log.:. \n")
            while not self.channel.recv_ready():
                pass
            out = self.channel.recv(9999)
            print(out)

    '''
    gets the current session directory
    '''

    def _pwd(self):
        stdin, stdout, stderr = self.exec_command(self.Commands.PWD)
        lines = stdout.readlines()
        return lines

    '''
    Setups the current session directory to home
    '''

    def _go_home(self):
        self.channel.send(self.Commands.GO_HOME+'\n')
        # while not self.channel.recv_ready():
        #         pass
        # out = self.channel.recv(9999)
        # print(out.decode())


    '''
    Setups the libraries path in SLM
    '''

    def _set_library_path(self):
        self.channel.send(self.Commands.SET_LIBRARY+'\n')
        # while not self.channel.recv_ready():
        #         pass
        # out = self.channel.recv(9999)
        # print(out.decode())

    '''
    Runs the show command on slm
    '''

    def _show_image(self):
        self.channel.send(self.Commands.SHOW_IMAGE+'\n')
        # while not self.channel.recv_ready():
        #         pass
        # out = self.channel.recv(9999)
        # print(out.decode())

    '''
    Validates the array whether it is a legitimate input.
    '''

    def _validateArray(self, array):
        if (array.shape == (self.height, self.width)) and (array.max() <= self.max) and (array.min() >= self.min):
            return array
        raise Exception(self.invalid_input_err)

    '''
    Validates the input image path whether it is a legitimate input. And returns the array in case it's a correct format
    '''

    def _validateImage(self, path):
        im = imageio.imread(path)
        return self._validateArray(im)

    '''Saves the image in the temporary path'''

    def _saveImage(self, array):
        try:
            imageio.imwrite(self.CACHE_PATH+'/tmp.bmp', array)
        except Exception as ex:
            print(ex)

    '''Removes the image in the temporary path'''

    def _removeImage(self, array):
        try:
            if os.path.exists(self.CACHE_PATH+'/tmp.bmp'):
                os.remove(self.CACHE_PATH+'/tmp.bmp')
        except Exception as ex:
            print(ex)

    '''
    Establishes connection to the Holoeyes SLM device
    '''

    def connect(self):
        return super().connect(self.hostname, port=self.port, username=self.username, password=self.password)

    '''
    Disoconnects the HDMI from the HoloeyeSLM device
    '''

    def diconnectHDMI(self):
        self.channel.send(self.Commands.DISABLE_HDMI+'\n')
        # while not self.channel.recv_ready():
        #         pass
        # out = self.channel.recv(9999)
        # print(out.decode())

    '''
    Sends an image at a given path to HoloeyeSLM
    '''

    def sendImage(self, FILE_PATH):
        try:
            im = imageio.imread(FILE_PATH)
            array = self._validateImage(FILE_PATH)
            self._saveImage(array)
            # TODO : SEND IMAGE TO HOME DIRECTORY
            with SCPClient(self.get_transport()) as scp:
                scp.put(self.CACHE_PATH+'/tmp.bmp',remote_path='~')
                scp.close()
            self._show_image()
            # TODO : RUN SHOW COMMAND ON

            # self._removeImage()
        except Exception as ex:
            raise ex
    '''
    Sends an nparray to HoloeyeSLM
    '''
    def sendData(self, data):
            try:
                array = self._validateArray(data)
                self._saveImage(array)
                # TODO : SEND IMAGE TO HOME DIRECTORY
                with SCPClient(self.get_transport()) as scp:
                    scp.put(self.CACHE_PATH+'/tmp.bmp',remote_path='~')
                    scp.close()
                self._show_image()
                # TODO : RUN SHOW COMMAND ON
                self._removeImage()
            except Exception as ex:
                raise ex

    '''
    Changes the current HoloeyeSLM IP address
    '''
    def changeIP(self, new_IP='10.10.70.2',mask='255.0.0.0'):
        print(self.Commands.CHANGE_IP.format(new_IP,mask)+'\n')
        self.channel.send(self.Commands.CHANGE_IP.format(new_IP,mask)+'\n')
        self.hostname = new_IP
        self.close()